frame/examples/kitchensink/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub foo: u32,
+		pub bar: T::BlockNumber,
+	}
+
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self { foo: 0, bar: Default::default() }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			Foo::<T>::put(self.foo);
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::set_foo_benchmark())]
+		pub fn set_foo(
+			_: OriginFor<T>,
+			new_foo: u32,
+			#[pallet::compact] _other_compact: u128,
+		) -> DispatchResult {
+			Foo::<T>::set(Some(new_foo));
+
+			Ok(())
+		}
+	}
+
+	/// The event type. This exactly like a normal Rust enum.
+	///
+	/// It can or cannot be generic over `<T: Config>`. Note that unlike a normal enum, if none of
+	/// the variants actually use `<T: Config>`, the macro will generate a hidden `PhantomData`
+	/// variant.
+	///
+	/// The `generate_deposit` macro generates a function on `Pallet` called `deposit_event` which
+	/// will properly convert the error type of your pallet into `RuntimeEvent` (recall `type
+	/// RuntimeEvent: From<Event<Self>>`, so it can be converted) and deposit it via
+	/// `frame_system::Pallet::deposit_event`.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A simple tuple style variant.
+		SomethingHappened(u32),
+		/// A simple struct-style variant. Note that we use `AccountId` from `T` because `T:
+		/// Config`, which by extension implies `T: frame_system::Config`.
+		SomethingDetailedHappened {
+			at: u32,
+			to: T::AccountId,
+		},
+		// /// Another variant.
+		SomeoneJoined(T::AccountId),
+	}
+
+	/// The error enum. Must always be generic over `<T>`, which is expanded to `<T: Config>`.
+	#[pallet::error]
+	pub enum Error<T> {
+		SomethingWentWrong,
+		SomethingBroke,
+	}
+
+	/// All the possible hooks that a pallet can have. See [`frame_support::traits::Hooks`] for more
+	/// info.
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn integrity_test() {}
+
+		fn offchain_worker(_n: T::BlockNumber) {
+			unimplemented!()
+		}
+
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			unimplemented!()
+		}
+
+		fn on_finalize(_n: T::BlockNumber) {
+			unimplemented!()
+		}
+
+		fn on_idle(_n: T::BlockNumber, _remaining_weight: Weight) -> Weight {
+			unimplemented!()
+		}
+
+		fn on_runtime_upgrade() -> Weight {
+			unimplemented!()
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
+			unimplemented!()
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade(_state: Vec<u8>) -> Result<(), TryRuntimeError> {
+			unimplemented!()
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn try_state(_n: T::BlockNumber) -> Result<(), TryRuntimeError> {
+			unimplemented!()
+		}
+	}
+
+	#[pallet::composite_enum]
```

Comment:
not everything is documented evenly in this crate. 

I actually think this crate should not document all pallet macros, but should only provide a very high level intro, and instead make sure when you hover over each pallet macro, the documentation is excellent. 

I have already started this for storage macros in https://github.com/paritytech/substrate/pull/13987. You can follow the same path.